### DISCUSSION TITLE
Express schema to database

### DISCUSSION THREAD

**Discussion by bryanveler** [IfcOpenShell/IfcOpenShell#4155] [Q&A] [answered]
I am trying to figure out which part of the ifcopenshell package creates tables and relationships that the express schema defines. I think I have found where the parsing happens and where it connects to a database but I'm trying to understand where in the package there is code that says to write to the database instead of memory. I feel like there should be something in the package that basically creates a database schema/template with all the tables and relationships, etc based off of the express schema, but I am not having any luck finding it. I'm just a construction contractor with an interest in development.  Any help will be greatly appreciated! 🙏

**Comment by aothms** [comment, marked as answer]
IfcOpenShell is not built on top of a database, but essentially just a couple of in memory maps to resolve instance references. If that's your thing, in the wider community there are initiatives to map to modern database backends https://github.com/Krande/ifcdb/ and we also have some mapping to sql in python https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/ifcopenshell/sql.py 

The combination of inheritance, ordered aggregate, unions (select types in EXPRESS), makes a 1 to 1 idiomatic mapping to a relational db extremely challenging. So that's why Krande/ifcdb uses EdgeDB and the SQL mapping in IfcOpenShell still needs some alignment to find the best approach.

**Reply by bryanveler** [reply]
Gotcha.  EdgeDB looks pretty interesting.  I'm going to look  into that more.  For  now though, to get  my head  wrapped around  everything  else  in the IfcOpenShell package and start playing with it, I think maybe i could get  some of  the functionality i am looking for by creating my own library of materials, products, costs, etc and just import it at the beginning of a  new file, correct?  If I change the cost for, say, a  2x4 wood stud in that  library, is there a  method in IfcOpenShell that i can call that would change the cost of a  WallElementedCase that uses the  2x4, accordingly?  And further, is there  a  method to update an existing ifc file per changes in the library that the file is using?  Thanks for your help!
